### DISCUSSION TITLE
Leaky abstraction on typescript types generation for apollo project.

### DISCUSSION THREAD

**Discussion by JustFly1984** [dotansimha/graphql-code-generator#6183] [Q&A]
I'm working on apollo server app. I have a schema from another app written in Ruby, and using types generated with `graphql-code-generator`

a graphql type, for example:

```
type Page {
 id: String!
 title: String!
 name: String
 items: [String!]!
 paragraphs: [Paragraph!]
 images: [Image]
}
```

results in typescript type declaration which looks like this:

```
  interface Page {
    id: string
    title: string
    name?: string | null
    items: string[]
    paragraphs?: {
      // not relevant
    }[] | null
    images?: ({
     // not relevant
    } | null)[] | null
  }
```

Two issues I see here is over-fetch data from DataBase, for fields not requested by the client.

The solution is pretty simple: It requires to wrap imported type in `Partial` utility type, to not , do not make joins in SQL, etc.

```
type PartialPage = Partial<Page>
```

and that is pretty much equal to

```
interface PartialPage {
  id?: string
  title?: string
  name?: string | null
  items?: string[]
  paragraphs?: {
    // not relevant
  }[] | null
  images?: ({
   // not relevant
  } | null)[] | null
}
```

But here is a second issue: you can't replace values with async functions, which should be possible due to existing apollo server functionality, to escape over-fetching in query and mutation function resolvers.

Am I missing something? Should type generation to be improved for apollo projects?

**Comment by JustFly1984** [comment]
I also had to change generated types

before:
```
export type ResolverTypeWrapper<T> = Promise<T> | T
```

after:
```
export type ResolverTypeWrapper<T> = Promise<Partial<T>> | Partial<T>
```

still thinking how to change types to add support for async functions as props in `T` for top level queries and mutations

**Reply by ardatan** [reply]
If I understand your concern properly; this is mostly user specific because the return type of resolver function depends on your setup. Since "Return type of resolver" !== "Return type of GraphQL Field", we have `mappers` and `defaultMapper` configuration options.
You can check this out also;
https://the-guild.dev/blog/better-type-safety-for-resolvers-with-graphql-codegen

**Reply by JustFly1984** [reply]
Is there a way to allow async functions? I've used `Partial<T>` and it does works.
Can I use something like `(Partial<T> | Promise<Partial<T>>)`?

**Reply by ardatan** [reply]
@JustFly1984 I think it should

**Reply by JustFly1984** [reply]
@ardatan I've tried to use defaultMapper with `(Partial<PageItemGuidance> | Promise<Partial<{T}>>)`

and generated types resulted in `PageItemGuidance: ResolverTypeWrapper<(Partial<PageItemGuidance> | Promise<Partial<{T}>>)>;`. Looks like replacement happens only for first `{T}`. :(
Would love to see it fixed.

**Reply by dotansimha** [reply]
@JustFly1984 I fixed it in: https://github.com/dotansimha/graphql-code-generator/pull/6307

**Reply by dotansimha** [reply]
Fixed in `@graphql-codegen/typescript-resolvers@1.20.0`.

**Reply by JustFly1984** [reply]
@dotansimha Actually today I had implemented correct type:

```ts
export type PartialAsync<T> = Partial<
  {
    [K in keyof T]: T[K] extends infer R ? R | (() => Promise<R>) : T
  }
>
```

but I'm not sure how to provide PartialAsync type for generator. I'm importing the same PartialAsync for using in return values of every mutation function.

In this case I could use `defaultMapper PartialAsync<{T}>`

meanwhile I'm editing autogenerated types manually.

While I'm editing, I see that some Scalar types get wrapped in Partial. `Partial<Scalar['Json']>` for example and several others. I think that it is incorrect typing generation.

This type allows me to return async nullable values from mutation functions in `apollo-server` project.

**Reply by JustFly1984** [reply]
@dotansimha How do I import PartialAsync type into generated types, to use it in defaultMapper ?

**Reply by dotansimha** [reply]
You can either specify it as mapper syntax: `./utils#PartialAsync` or, you can use `add` plugin to add the import you need.
